### PR TITLE
armin: top-3 checkpoint ensemble (post-processing gain)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

After training completes, a single best-validation checkpoint is used for test inference. However, checkpoints from adjacent good epochs lie in nearby but distinct minima in weight space. Averaging their **predictions** (not weights) over multiple checkpoints gives a free ensemble — each checkpoint makes slightly different errors, and averaging cancels independent noise components.

This is distinct from EMA (which averages weights during training) and from test-time augmentation. **Checkpoint ensemble** averages the *outputs* of independently saved EMA checkpoints from the best 3 epochs, at test time only — no additional training, no code change to train.py beyond saving additional checkpoints.

Expected gains: 1–3% reduction in abupt_axis_mean based on ensemble literature (standard bias-variance tradeoff: multiple models at different weight positions in the same basin reduce variance). Especially useful for the noisy tau_y/tau_z axes which have high point-to-point variance.

**Why this might be uniquely effective for CFD wall-shear:** The model's prediction at each epoch is biased by which car geometries it happened to overfit slightly. A 3-checkpoint ensemble effectively averages over 3 slightly different "biases" in the val-selected checkpoint, giving a more robust prediction across the test geometries.

## Instructions

**Step 1 — Modify train.py to save top-K EMA checkpoints** (not just the single best):

Add `--save-top-k-checkpoints N` (int, default `1`) to `Config` and argparser. When `N > 1`, maintain a sorted list of the top-N validation checkpoints by val_primary metric and save each one as `checkpoint_top{rank}_epoch{ep}.pt`.

```python
# In the validation checkpoint save logic:
if config.save_top_k_checkpoints > 1:
    topk_checkpoints.append((val_abupt, epoch, ema_state_dict_copy))
    topk_checkpoints.sort(key=lambda x: x[0])  # ascending (lower is better)
    topk_checkpoints = topk_checkpoints[:config.save_top_k_checkpoints]
    for rank, (metric, ep, state) in enumerate(topk_checkpoints):
        path = f"{config.checkpoint_dir}/checkpoint_top{rank+1}_epoch{ep}.pt"
        torch.save(state, path)
```

**Step 2 — Add ensemble inference script** `target/ensemble_eval.py`:

```python
"""
Ensemble multiple EMA checkpoints and evaluate on val+test splits.
Usage:
    python ensemble_eval.py \
        --checkpoints checkpoint_top1_epoch3.pt checkpoint_top2_epoch4.pt checkpoint_top3_epoch2.pt \
        --config-from-run <wandb_run_id>  # to recover the exact Config used for training
"""
import torch, argparse
from train import Config, build_model, evaluate  # reuse existing code

def load_model(checkpoint_path, config):
    model = build_model(config)
    state = torch.load(checkpoint_path, map_location='cuda')
    model.load_state_dict(state)
    model.eval()
    return model

def ensemble_predict(models, batch):
    """Average predictions from multiple models."""
    with torch.no_grad():
        preds = [model(batch) for model in models]
    # preds is list of (surf_pred, vol_pred) tuples
    surf_avg = torch.stack([p[0] for p in preds]).mean(0)
    vol_avg = torch.stack([p[1] for p in preds]).mean(0)
    return surf_avg, vol_avg
```

**Step 3 — Run the full experiment in two phases:**

**Phase 1 (training):** Run the PR #99 base config with `--save-top-k-checkpoints 3` to save the top 3 checkpoints:

```bash
cd target/
python train.py \
  --save-top-k-checkpoints 3 \
  --volume-loss-weight 2.0 \
  --batch-size 8 \
  --validation-every 1 \
  --lr 5e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 \
  --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 \
  --wallshear-z-weight 2.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group armin-checkpoint-ensemble-r10 \
  --wandb-run-name armin-ckpt-train
```

**Phase 2 (ensemble eval):** After training, evaluate 3 ensemble configurations:

| Config | Checkpoints | Notes |
|---|---|---|
| Single best (control) | top1 only | Matches standard eval |
| Top-2 ensemble | top1 + top2 | 2-model average |
| Top-3 ensemble | top1 + top2 + top3 | 3-model average |

Report `val_primary/*` and `test_primary/*` for all 3 configs. The key question: does ensemble reduce test abupt below 11.73?

**What to report in results:**
1. Training run val abupt trajectory and best single-checkpoint metrics (same as any PR)
2. Ensemble val/test metrics for top-2 and top-3 vs single-checkpoint
3. Per-axis breakdown: does ensemble help tau_y/tau_z more than tau_x? (expected yes — higher variance axes benefit more from averaging)

## Baseline (PR #99, W&B run `3hljb0mg`)

| Metric | yi best (val, single ckpt) | yi best (test, single ckpt) | AB-UPT target |
|---|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **10.69** | **11.73** | — |
| `surface_pressure_rel_l2_pct` | 6.97 | 6.64 | 3.82 |
| `wall_shear_rel_l2_pct` | 11.69 | 11.48 | 7.29 |
| `volume_pressure_rel_l2_pct` | 7.85 | 14.42 | 6.08 |
| `wall_shear_x_rel_l2_pct` | 10.17 | 10.06 | 5.35 |
| `wall_shear_y_rel_l2_pct` | **13.73** | **13.53** | **3.65** |
| `wall_shear_z_rel_l2_pct` | **14.73** | **13.98** | **3.63** |

Reproduce baseline:
```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 5e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #155 because ADVISOR: Closing as orphaned. No senpai-yi-armin pod exists on this track and 'armin' is not in the active student roster, so this PR will never be picked up by an agent. Checkpoint ensembling remains a viable direction and can be reassigned to an active student in a future round..
